MatrixElement gets enhanced free symbols

smichr opened the pull request:
#### Brief description of what is fixed or changed

This makes MatrixElement more like Indexed in that
it returns itself as well as free symbols of the object.
In addition,

```python
>>> IndexedBase('x')[0].free_symbols
{x[0], x}
>>> MatrixSymbol('x',2,2)[0,0].free_symbols
{x, x[0, 0]}
```
#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
    * MatrixElement now lists symbolic indices and the MatrixSymbol as free symbols
<!-- END RELEASE NOTES -->

smichr commented on sympy/matrices/expressions/tests/test_matexpr.py:
```diff
@@ -156,8 +160,9 @@ def test_dense_conversion():
-def test_free_symbols():
+def test_matexpr_free_symbols():
     assert (C*D).free_symbols == {C, D}
+    assert C[x, 0].free_symbols == {C, x, C[x, 0]}
```
substitutions now fail if the dimensions are incompatible
```python
>>> X=MatrixSymbol("X",2,2)
>>> X[y,1]
X[y, 1]
>>> _.subs(y,1)
X[1, 1]
>>> X[y,1].subs(X,MatrixSymbol('Z',3,3))
X[y, 1]  <-- unchanged
>>> X[y,1].subs(X,MatrixSymbol('Z',1,1))
X[y, 1]  <-- unchanged
```

ThePauliPrinciple replied:
Substituting in `'Z',3,3` for `'X',2,2` seems compatible to me since it is bigger than `X`

smichr commented:
If there are no objections, I will commit this in 24h.

ThePauliPrinciple commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -642,6 +655,17 @@ def _eval_derivative(self, v):
+        # when replacing the symbol, the indices cannot
+        # access out of bounds
+        if old == self.symbol:
+            if old.func == new.func:
```
Do you really want to compare `func` here, or are you actually looking for `type(old) == type(new)`?

smichr replied:
I don't have a nuanced understanding considering this:
```python
>>> Dummy().func
<class 'sympy.core.symbol.Dummy'>
>>> type(Dummy())
<class 'sympy.core.symbol.Dummy'>
```
So for me these seem like the same thing.

ThePauliPrinciple replied:
Currently, `func` always returns the type, however, that means that `atom.func(*atom.args) == atom` generally can't work for `Atom`s (unless the atom stores no information) if `.func` always returns its type since it never stores it in `args`. So it should be made possible that `func` can also return something else.

I think the use of `func` as an alias for `type()` like this is more a remainder of python 2 style classes rather than something that should generally be prefered. If the question is whether it is the same class, then `type()` should be used since it is clearer to the general python reader and it would allow for a more flexible use of `func`.

Maybe @oscarbenjamin has some further insight.

smichr commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -642,6 +655,17 @@ def _eval_derivative(self, v):
+        # when replacing the symbol, the indices cannot
+        # access out of bounds
+        if old == self.symbol:
+            if old.func == new.func:
```
```suggestion
            if type(old) == type(new):
```

oscarbenjamin commented:
I'm not sure I understand the purpose of this change. Now that we have `has_free` do we need e.g. `M[1, 1]` to be in its own free symbols?

smichr commented:
One can differentiate wrt a free symbol so it makes sense to show in a query for free symbols those items that have that quality:
```python
>>> a=MatrixSymbol('A',2,2)
>>> a.diff(a)
PermuteDims(ArrayTensorProduct(I, I), (3)(1 2))
>>> eq=a[0,0]*2
>>> eq.diff(a[0,0])
2
>>> eq = 2*x + 1
>>> eq.diff(x)
2
```
`has`/`has_free` simply tell whether a given node or subset of args for Add/Mul have a given expression.
```python
>>> eq=2*x + 1
>>> eq.has(2*x)
True
>>> eq.has(x)
True
>>> eq.has(2*x + 1)
True
>>> eq.has(2*x + 2)
False
```
But "having" something doesn't indicate whether that thing is symbol like:
```
>>> eq.diff(2*x)
Traceback (most recent call last):
...
ValueError:
Can't calculate derivative wrt 2*x.
```

smichr committed:
MatrixElement gets name and symbol properties

smichr committed:
check indices

smichr committed:
type: ignore

in response to these errors

ympy/tensor/array/expressions/conv_array_to_matrix.py:288: error: No overload variant of "__getitem__" of "tuple" matches argument type "List[Any]"
sympy/tensor/array/expressions/conv_array_to_matrix.py:288: note: Possible overload variants:
sympy/tensor/array/expressions/conv_array_to_matrix.py:288: note:     def __getitem__(self, int) -> Basic
sympy/tensor/array/expressions/conv_array_to_matrix.py:288: note:     def __getitem__(self, slice) -> Tuple[Basic, ...]
sympy/tensor/array/expressions/conv_array_to_matrix.py:290: error: No overload variant of "__getitem__" of "tuple" matches argument type "List[Any]"
sympy/tensor/array/expressions/conv_array_to_matrix.py:290: note: Possible overload variants:
sympy/tensor/array/expressions/conv_array_to_matrix.py:290: note:     def __getitem__(self, int) -> Basic
sympy/tensor/array/expressions/conv_array_to_matrix.py:290: note:     def __getitem__(self, slice) -> Tuple[Basic, ...]
sympy/tensor/array/expressions/conv_array_to_matrix.py:494: error: Incompatible types in assignment (expression has type "List[Tuple[Any, ...]]", variable has type "Dict[Any, Tuple[Any, ...]]")
sympy/tensor/array/expressions/conv_array_to_matrix.py:500: error: Incompatible types in assignment (expression has type "List[Any]", variable has type "Dict[Any, Tuple[Any, ...]]")
sympy/tensor/array/expressions/arrayexpr_derivatives.py:26: error: "Expr" has no attribute "shape"
sympy/tensor/array/expressions/arrayexpr_derivatives.py:66: error: "Expr" has no attribute "shape"
sympy/tensor/array/expressions/arrayexpr_derivatives.py:77: error: "Expr" has no attribute "shape"
sympy/tensor/array/expressions/arrayexpr_derivatives.py:82: error: "Expr" has no attribute "shape"

smichr committed:
add free symbols and update codegen

smichr committed:
no Trace in hyper_sum; conv mut to immut

smichr commented:
This is the wrong direction to be going. Indexed should not list itself as a free symbol; we should not make MatrixElement behave the same way.
